cpp/ql/src/Security/CWE/CWE-295/SSLResultNotChecked.ql

```diff
@@ -0,0 +1,122 @@
+/**
+ * @name Certificate not checked
+ * @description Always check the result of certificate verification after fetching an SSL certificate.
+ * @kind problem
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision medium
+ * @id cpp/certificate-not-checked
+ * @tags security
+ *       external/cwe/cwe-295
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+import semmle.code.cpp.controlflow.IRGuards
+
+/**
+ * A call to `SSL_get_peer_certificate`.
+ */
+class SSLGetPeerCertificateCall extends FunctionCall {
+  SSLGetPeerCertificateCall() {
+    getTarget().getName() = "SSL_get_peer_certificate" // SSL_get_peer_certificate(ssl)
+  }
+
+  Expr getSSLArgument() { result = getArgument(0) }
+}
+
+/**
+ * A call to `SSL_get_verify_result`.
+ */
+class SSLGetVerifyResultCall extends FunctionCall {
+  SSLGetVerifyResultCall() {
+    getTarget().getName() = "SSL_get_verify_result" // SSL_get_peer_certificate(ssl)
+  }
+
+  Expr getSSLArgument() { result = getArgument(0) }
+}
+
+/**
+ * Holds if the SSL object passed into `SSL_get_peer_certificate` is checked with
+ * `SSL_get_verify_result` entering `node`.
+ */
+predicate resultIsChecked(SSLGetPeerCertificateCall getCertCall, ControlFlowNode node) {
+  exists(Expr ssl, SSLGetVerifyResultCall check |
+    ssl = globalValueNumber(getCertCall.getSSLArgument()).getAnExpr() and
+    ssl = check.getSSLArgument() and
+    node = check
+  )
+}
+
+/**
+ * Holds if the certificate returned by `SSL_get_peer_certificate` is found to be
+ * `0` on the edge `node1` to `node2`.
+ */
+predicate certIsZero(
+  SSLGetPeerCertificateCall getCertCall, ControlFlowNode node1, ControlFlowNode node2
+) {
+  exists(GuardCondition guard, Expr cert |
+    cert = globalValueNumber(getCertCall).getAnExpr() and
+    (
+      exists(Expr zero |
+        zero.getValue().toInt() = 0 and
+        node1 = guard and
+        (
+          // if (cert == zero) {
+          guard.comparesEq(cert, zero, 0, true, true) and
+          node2 = guard.getATrueSuccessor()
+          or
+          // if (cert != zero) { }
+          guard.comparesEq(cert, zero, 0, false, true) and
+          node2 = guard.getAFalseSuccessor()
+        )
+      )
+      or
+      // if (cert) { }
+      guard = cert and
+      node1 = guard and
+      node2 = guard.getAFalseSuccessor()
+      or
+      // if (!cert) {
+      node1 = guard.getParent() and
+      node2 = guard.getParent().(NotExpr).getATrueSuccessor()
```

Comment:
Could these cases not be covered by using `guard.controls`? That might be able to solve the FP in `test2_11` (and in principle the one in `test2_9` as well, but IIRC the `GuardCondition`s class specifically doesn't handle this case very well).

Reply:
I've pushed an improved version using `guard.controls` (and fixing some other issues).  It's a bit tricky because guards gives the basic blocks that may be entered _only_ if the guard goes a particular way, whereas we want to be able to handle blocks that have more than one route in, e.g.
```
cert = SSL_get_peer_certificate(ssl);
if (cert)
{
  result = SSL_get_verify_result(ssl);
  ...
}
// here
```
In the case above `here` is good because one path to it has `cert = 0`, the other has a call to `SSL_get_verify_result` - not because of either fact on its own.  As such the code is still somewhat special-cased rather than just letting guards do its thing.